v1/contracts/FixedPoint.sol

```diff
@@ -0,0 +1,51 @@
+pragma solidity ^0.5.0;
+
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+
+/**
+ * @title Library for fixed point arithmetic on uints
+ * TODO(ptare): Add operations against non-fixed point number, e.g., add(Unsigned, uint).
+ */
+library FixedPoint {
+
+    using SafeMath for uint;
+
+    // Supports 18 decimals. E.g., 1e18 represents "1", 5e17 represents "0.5".
+    // Can represent a value up to (2^256 - 1)/10^18 = ~10^59.
+    uint private constant FP_SCALING_FACTOR = 10**18;
+
+    struct Unsigned {
+        uint value;
+    }
+
+    /** @dev Adds two `Unsigned`s, reverting on overflow. */
+    function add(Unsigned memory a, Unsigned memory b) internal pure returns (Unsigned memory) {
+        return Unsigned(a.value.add(b.value));
+    }
+
+    /** @dev Subtracts two `Unsigned`s, reverting on underflow. */
+    function sub(Unsigned memory a, Unsigned memory b) internal pure returns (Unsigned memory) {
+        return Unsigned(a.value.sub(b.value));
+    }
+
+    /** @dev Multiplies two `Unsigned`s, reverting on overflow. */
+    function mul(Unsigned memory a, Unsigned memory b) internal pure returns (Unsigned memory) {
+        // There are two caveats with this computation:
+        // 1. Max output is ~10^41 (vs ~10^59 as the max value that can be represented), otherwise an intermediate value
```

Comment:
Is this correct? I thought the max value that could be represented was ~10^71 and therefore the max output of this would be ~10^53?

Reply:
I might be understanding this incorrectly, but here's what I think:

Fixed point in binary:
The lowest 18 bits are below the binary point, the upper 238 bits are above. Then, 2^238-1 = ~10^71 is the max representable number.

Fixed point in decimal:
18 digits after the decimal, ~59 bits represent the decimal and the upper ~197 bits represent the whole number. The max would be (2^256-1)/10^18 = ~10^59.

Reply:
Oh! I see - I was thinking you were talking about the raw number, not the FixedPoint interpretation of the number. That definitely makes sense. It might make sense to somehow note that the limits are on the fixed point interpretation, not the raw `uint`. Or maybe separate it into right of the decimal and left of the decimal portions.

Reply:
I expanded on the comments a bit, let me know how else we can improve them.

This discussion raised a question regarding the behavior of the mixed functions like `add(Unsigned a, uint b)`. Is `b` already scaled up by 10**18? E.g., if the caller passes `b=10**17`, is that interpreted as `0.1` or as `10**17`.

Reply:
I think `b` would be interpreted as a normal integer - meaning `10^17` would just map to `10^17` rather than `0.1`. If they wanted it to be recognized as fixed point, they should have used a fixed point type.

IMO, our philosophy from this point forward should be that if a number isn't wrapped in `FixedPoint.Unsigned` or `FixedPoint.Signed`, it should be interpreted as the same scale as the underlying integer. Is that what you were thinking?

Reply:
And your expanded comment looks awesome - can you also add that same language to the `.div()`?

Reply:
Oops forgot to update the other comment, done.

That's exactly how I was thinking the mixed operations should work.